spinalcordtoolbox/reports/qc2.py

```diff
@@ -204,263 +205,300 @@ def rescale_grid(
+        slice_labels = [str(i) for i in range(num_slices)]
+        if "L" in img.orientation:
+            slice_labels.reverse()
+
+        return SlicingSpec(
+            affine=affine,
+            offsets={
+                label: np.array([i, 0, 0], dtype=np.float64)
+                for i, label in enumerate(slice_labels)
+            },
+            shape=shape,
+            axis_labels=(("S", "I"), ("P", "A")),
+        )
+
+    def center_columns(self, seg: Image) -> 'SlicingSpec':
+        """
+        A re-centered and cropped slicing spec, meant for sagittal views.
+
+        Each slice is individually centered in the last axis around the center
+        of mass of the given segmentation. It is also cropped in the last axis
+        to the larger of:
+        - 110% of the maximum width of the segmentation, or
+        - 50% of the total image.
+
+        Any slices that are more than 2 slices away from the segmentation are
+        also dropped.
+        """
+        # Check which slices contain the segmentation, and compute their width.
+        slices_seg = self.get_slices(seg, order=1)
+        width = {}
+        for i, slice_seg in enumerate(slices_seg.values()):
+            [columns] = slice_seg.any(axis=0).nonzero()
+            if columns.size != 0:
+                width[i] = int(columns.max()) - int(columns.min()) + 1
+        if not width:
+            raise ValueError("The mask image is empty. Cannot crop using an empty mask. Check the input (e.g. '-qc-seg').")
+
+        # Compute the target slice width.
+        shape = (
+            self.shape[0],  # the full image height
+            math.ceil(max(
+                1.1 * max(width.values()),  # 110% of the segmentation width
+                0.5 * self.shape[1],  # 50% of the full image width
+            )),
+        )
+
+        # Drop slices that are far from the segmentation.
+        slice_min = max(0, min(width.keys()) - 2)
+        slice_max = min(len(slices_seg) - 1, max(width.keys()) + 2)
+        offsets_dict = dict(list(self.offsets.items())[slice_min:slice_max+1])
+
+        # Re-center each remaining slice.
+        for slice_label, offset in offsets_dict.items():
+            offset[2] += ndimage.center_of_mass(slices_seg[slice_label])[1]
+
+        # Take into account the cropping width.
+        offsets_array = np.array(list(offsets_dict.values()))
+        inf_nan_fill(offsets_array[:, 2])
+        offsets_array[:, 2] -= (shape[1] - 1) / 2
+        offsets_dict = dict(zip(offsets_dict.keys(), offsets_array))
+
+        return SlicingSpec(
+            affine=self.affine,
+            offsets=offsets_dict,  # new offsets
+            shape=shape,  # new shape
+            axis_labels=self.axis_labels,
+        )
+
+    def generate_mosaic(self, img: Image, path: Path, scale: float = 2.5):
+        """
+        Extract slices from an image and save the result as a png.
+        """
```

Comment:
Do we want to add some input validation here to ensure that `slicing_spec` has its properties set before proceeding?

Reply:
I'll note that all the properties of `SlicingSpec` are required arguments to its `__init__` method, because none of them has a default value assigned at the class level. So, this is already somewhat validated.